File: DelSole.CrossAnalyzers.VisualBasic/DelSole.CrossAnalyzers.VisualBasic/DateTimeOffsetAnalyzer.vb
Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

<DiagnosticAnalyzer(LanguageNames.VisualBasic)>
Public Class DateTimeOffsetAnalyzer
    Inherits DiagnosticAnalyzer

    Public Const DiagnosticId = "CRA013"
    Friend Shared ReadOnly Title As LocalizableString = "Assigning DateTime to objects of type DateTimeOffset might result in runtime errors"
    Friend Shared ReadOnly MessageFormat As LocalizableString = "'{0}' should be of type DateTimeOffset"
    Friend Const Category = "Platform"

    Friend Shared Rule As New DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, True, helpLinkUri:="https://github.com/AlessandroDelSole/DelSole.CrossAnalyzers/wiki")

    Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
        Get
            Return ImmutableArray.Create(Rule)
        End Get
    End Property

    Public Overrides Sub Initialize(context As AnalysisContext)
        context.RegisterSyntaxNodeAction(AddressOf AnalyzeAssignementStatements, SyntaxKind.SimpleAssignmentStatement)
    End Sub

    Public Sub AnalyzeAssignementStatements(context As SyntaxNodeAnalysisContext)
        Dim node = TryCast(context.Node, AssignmentStatementSyntax)
        If node Is Nothing Then
            Return
        End If


        Dim leftSymbolInfo = context.SemanticModel.GetSymbolInfo(node.Left).Symbol

        If leftSymbolInfo.Kind = SymbolKind.Local Then Return

        Dim tempSymbol As Boolean

        If leftSymbolInfo.Kind = SymbolKind.Property Then
            tempSymbol = CType(leftSymbolInfo, IPropertySymbol).Type.ToString().Contains("DateTimeOffset")

        Else
            tempSymbol = CType(leftSymbolInfo, IFieldSymbol).Type.ToString().Contains("DateTimeOffset")
        End If

        If tempSymbol = False Then Return

        Dim returnType = ""
        Dim SymbolInfo = context.SemanticModel.GetSymbolInfo(node.Right).Symbol
        If TypeOf (SymbolInfo) Is ILocalSymbol Then
            returnType = CType(SymbolInfo, ILocalSymbol).Type.ToString()
        ElseIf TypeOf (SymbolInfo) Is IPropertySymbol Then
            returnType = CType(SymbolInfo, IPropertySymbol).Type.ToString()
        ElseIf TypeOf (SymbolInfo) Is IFieldSymbol Then
            returnType = CType(SymbolInfo, IFieldSymbol).Type.ToString()
        ElseIf TypeOf (SymbolInfo) Is IMethodSymbol Then
            returnType = CType(SymbolInfo, IMethodSymbol).ReturnType.ToString()
        Else
            Return
        End If

        If Not returnType.ToLowerInvariant() = "date" Then Return

        Dim diagn = Diagnostic.Create(Rule, node.Right.GetLocation(), node.Right.ToString())
        context.ReportDiagnostic(diagn)
    End Sub
End Class

File: DelSole.CrossAnalyzers.CSharp/DelSole.CrossAnalyzers.CSharp/DateTimeOffsetAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace DelSole.CrossAnalyzers.CSharp
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class DateTimeOffsetAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "CRA003";
        internal static readonly LocalizableString Title = "Assigning DateTime to objects of type DateTimeOffset might result in runtime errors";
        internal static readonly LocalizableString MessageFormat = "'{0}' should be of type DateTimeOffset";
        internal const string Category = "Platform";

        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, true);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxNodeAction(AnalyzeAssignementStatements, SyntaxKind.SimpleAssignmentExpression);

        }

        public void AnalyzeAssignementStatements(SyntaxNodeAnalysisContext context)
        {
            var node = context.Node as AssignmentExpressionSyntax;
            if (node == null)
            {
                return;
            }

            //Check if the node contains bindable properties
            ISymbol leftSymbolInfo = context.SemanticModel.GetSymbolInfo(node.Left).Symbol;

            if (leftSymbolInfo.Kind == SymbolKind.Local) return;

            bool tempSymbol;

            if (leftSymbolInfo.Kind==SymbolKind.Property)
            {
                tempSymbol = ((IPropertySymbol)leftSymbolInfo).Type.ToString().Contains("DateTimeOffset");
                
            }
            else
            {
                tempSymbol = ((IFieldSymbol)leftSymbolInfo).Type.ToString().Contains("DateTimeOffset");
            }

            if (tempSymbol == false) return;

            string returnType = "";
            var symbolInfo = context.SemanticModel.GetSymbolInfo(node.Right).Symbol;
            //Check if local variable
            if ((symbolInfo) is ILocalSymbol)
            {
                returnType = ((ILocalSymbol)symbolInfo).Type.ToString();
                //Check if property
            }
            else if ((symbolInfo) is IPropertySymbol)
            {
                returnType = ((IPropertySymbol)symbolInfo).Type.ToString();
                //Check if field
            }
            else if ((symbolInfo) is IFieldSymbol)
            {
                returnType = ((IFieldSymbol)symbolInfo).Type.ToString();
            }
            else if ((symbolInfo) is IMethodSymbol)
            {
                //Check if method
                returnType = ((IMethodSymbol)symbolInfo).ReturnType.ToString();
            }
            else
            {
                return;
            }

            if (returnType.ToLowerInvariant()=="system.datetime")
            {
                var diagn = Diagnostic.Create(Rule, node.Right.GetLocation(), node.Right.ToString());
                context.ReportDiagnostic(diagn);
            }
        }
    }
}